cmd/autoscaler/main.go

```diff
@@ -179,9 +179,13 @@ func scalerConfig(logger *zap.SugaredLogger) *autoscaler.DynamicConfig {
 
 func uniScalerFactoryFunc(endpointsInformer corev1informers.EndpointsInformer) func(decider *autoscaler.Decider, dynamicConfig *autoscaler.DynamicConfig) (autoscaler.UniScaler, error) {
 	return func(decider *autoscaler.Decider, dynamicConfig *autoscaler.DynamicConfig) (autoscaler.UniScaler, error) {
+		if decider.Labels == nil {
+			return nil, fmt.Errorf("no labels set on Decider: %v", decider)
+		}
+
 		// Create a stats reporter which tags statistics by PA namespace, configuration name, and PA name.
 		reporter, err := autoscaler.NewStatsReporter(decider.Namespace,
-			labelValueOrEmpty(decider, serving.ServiceLabelKey), labelValueOrEmpty(decider, serving.ConfigurationLabelKey), decider.Name)
+			decider.Labels[serving.ServiceLabelKey], decider.Labels[serving.ConfigurationLabelKey], decider.Name)
```

Comment:
do we actually care that those are empty? As far as I can tell we just pass them to the tag list.

Reply:
No, I think we're fine with them being empty.

Reply:
Then why do we fail when no labels are specified. Say labels are _there_, but non of them is what checked for?

Reply:
2 lines below this code we depend on a label being set (that code would've panicked today if Labels were nil)

Reply:
https://github.com/knative/serving/blob/8a6c8f921fa0748e27a4f94ccf6579def08afc34/cmd/autoscaler/main.go#L189-L192

Reply:
We do care about. For Prometheus, a tag should be consistent present for all data points of a metric. If a tag is not empty for some data points but empty for others, Prometheus will complain. Stackdriver has similar requirement for some labels.

Reply:
So, namespace, service, configuration and revision shouldn't be empty. In the cases that a revision is defined via configuration directly, we use "unknown" value instead of empty.